File: backend/src/main/java/at/ac/tuwien/sepm/assignment/individual/service/HorseService.java
package at.ac.tuwien.sepm.assignment.individual.service;

import at.ac.tuwien.sepm.assignment.individual.dto.HorseDataDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseQueryParamsDto;
import at.ac.tuwien.sepm.assignment.individual.entity.Horse;
import at.ac.tuwien.sepm.assignment.individual.exception.NotFoundException;
import at.ac.tuwien.sepm.assignment.individual.exception.ServiceException;
import at.ac.tuwien.sepm.assignment.individual.exception.StateConflictException;
import at.ac.tuwien.sepm.assignment.individual.exception.ValidationException;

import java.util.List;

/**
 * Service Object for Horses
 * Enables validation of incoming dtos and transferring the requests to the layers below.
 */
public interface HorseService {
    /**
     * Lists all horses stored in the system by search parameter qParams.
     *
     * @throws ValidationException if the argument is semantically incorrect
     * @throws StateConflictException if the argument conflicts with the state of the system
     * @throws ServiceException if an unexpected error in the layers below occurred
     *
     * @param qParams the query parameters of the search
     * @return list of all stored horses
     */
    List<Horse> allHorses(HorseQueryParamsDto qParams) throws ServiceException;

    /**
     * Creates a new horse in the system.
     *
     * @throws IllegalArgumentException it the argument is syntactically incorrect
     * @throws ValidationException if the argument is semantically incorrect
     * @throws StateConflictException if the argument conflicts with the state of the system
     * @throws ServiceException if an unexpected error in the layers below occurred
     *
     * @param dto the horse to create
     * @return the created horse
     */
    Horse createHorse(HorseDataDto dto) throws ServiceException;

    /**
     * Modifies an existing horse in the system.
     *
     * @throws NotFoundException if the horse with 'id' does not exist in the system
     * @throws IllegalArgumentException it the argument is syntactically incorrect
     * @throws ValidationException if the argument is semantically incorrect
     * @throws StateConflictException if the argument conflicts with the state of the system
     * @throws ServiceException if an unexpected error in the layers below occurred
     *
     * @param dto horse with the new data
     * @return the modified horse
     */
    Horse editHorse(long id, HorseDataDto dto) throws ServiceException;

    /**
     * Delete an existing horse in the system.
     *
     * @throws NotFoundException if the horse with 'id' does not exist in the system
     * @throws ValidationException if the argument is semantically incorrect
     * @throws ServiceException if an unexpected error in the layers below occurred
     *
     * @param id the horses id
     * @return the deleted horse
     */
    Horse deleteHorse(long id) throws ServiceException;

    /**
     * Retrieve a horse existing in the system by its id.
     *
     * @throws NotFoundException if the horse with 'id' does not exist in the system
     * @throws ValidationException if the argument is semantically incorrect
     * @throws ServiceException if an unexpected error in the layers below occurred
     *
     * @param id the horses id
     * @return the corresponding horse
     */
    Horse getHorse(long id) throws ServiceException;
}

File: frontend/src/app/service/horse.service.ts
import {HttpClient, HttpParams} from '@angular/common/http';
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {environment} from 'src/environments/environment';
import {Horse} from '../dto/horse';
import {HorseData} from '../dto/horseData';
import {HorseQuery} from '../dto/horseQuery';

const baseUri = environment.backendUrl + '/horses';

@Injectable({
  providedIn: 'root'
})
export class HorseService {

  constructor(
    private http: HttpClient,
  ) { }

  /**
   * Get all horses stored in the system by search parameter param
   *
   * @param params object that contains the search parameters
   * @return observable list of found horses.
   */
  getAll(params: HorseQuery): Observable<Horse[]> {
    let httpParams = new HttpParams();
    for (const key in params) {
      if (params[key]) {
        httpParams = httpParams.set(key, params[key]);
      }
    }
    return this.http.get<Horse[]>(baseUri, {params: httpParams});
  }


  /**
   * Create a new horse in the system
   *
   * @param horse the data of the horse to create
   * @return observable the created horse
   */
  createHorse(horse: HorseData): Observable<Horse> {
    return this.http.post<Horse>(baseUri, horse);
  }

  /**
   * Edit an already existing horse in the system.
   *
   * @param id the id of the horse to edit.
   * @param horse the updated values for the horse.
   * @return observable the edited horse
   */
  editHorse(id: string, horse: HorseData): Observable<Horse> {
    return this.http.put<Horse>(baseUri + '/' + id, horse);
  }

  /**
   * Get a horse stored in the system.
   *
   * @param id the id of the stored horse.
   * @return horse the horse with 'id'
   */
  getHorse(id: string) {
    return this.http.get<Horse>(baseUri + '/' + id);
  }

  /**
   * Delete a horse from the system.
   *
   * @param id the id of the horse to delete.
   * @return horse the deleted horse
   */
  deleteHorse(id: string) {
    return this.http.delete<Horse>(baseUri + '/' + id);
  }
}
